aws-greengrass-testing-features/aws-greengrass-testing-features-api/src/main/java/com/aws/greengrass/testing/features/SystemMetricSteps.java

```diff
@@ -0,0 +1,113 @@
+        this.cpu = hal.getProcessor();
+        this.ram = hal.getMemory();
+        this.scenarioContext = scenarioContext;
+    }
+
+    /**
+     * Record the current CPU or RAM usage depending on the step.
+     *
+     * @param stat the statistic to record, must be CPU or RAM
+     * @param statKey the key to store the statistic under in scenario context
+     * @throws IllegalArgumentException when parameter is not CPU or RAM
+     */
+    @When("I record the device's {word} usage statistic as {word}")
+    public void recordCpuOrRamInContext(String stat, String statKey) throws IllegalArgumentException {
+        switch (stat) {
+            case "CPU":
+                double cpuLoad = cpu.getSystemCpuLoad(500) * 100;
+                LOGGER.debug("System CPU load recorded as: " + cpuLoad + "%");
+                scenarioContext.put(statKey, Double.toString(cpuLoad));
+                break;
+            case "RAM":
+                double usedRam = (double) (ram.getTotal() - ram.getAvailable()) / (1024 * 1024);
+                LOGGER.debug("Used RAM recorded as: " + usedRam + " MB");
+                scenarioContext.put(statKey, Double.toString(usedRam));
+                break;
+            default:
+                throw new IllegalArgumentException("Please specify either CPU or RAM to track.");
+        }
+    }
+
+    /**
+     * Check difference between last two records and verify it is below a threshold.
+     *
+     * @param stat the statistic to record, must be CPU or RAM
+     * @param statKey1 the first key to check in scenarioContext, the one sampled earlier in time
+     * @param statKey2 the second key to check in scenarioContext, the one sampled later in time
+     * @param threshold the threshold to assert the difference is under, % for CPU and MB for RAM
+     * @param units dummy parameter so that the step is picked up, always % for CPU and MB for RAM
+     * @throws Exception when step fails due to exceeding the provided threshold or sample steps haven't run twice
+     * @throws IllegalArgumentException when stat param is not CPU or RAM
+     */
+    @Then("the difference between {word} usage samples {word} and {word} is less than {int} {word}")
+    public void checkSampleDiff(String stat,
+                                String statKey1,
+                                String statKey2,
+                                int threshold,
+                                String units) throws Exception, IllegalArgumentException {
+        switch (stat) {
+            case "CPU":
+                if (!units.equals("percent")) {
+                    throw new IllegalArgumentException("Please use percent as the unit of measurement for CPU usage.");
+                }
+                double cpuSample2 = Double.parseDouble(scenarioContext.get(statKey2));
+                double cpuSample1 = Double.parseDouble(scenarioContext.get(statKey1));
+                double cpuDiff = cpuSample2 - cpuSample1;
```

Comment:
Should this be absolute if we're checking only the difference? This is assuming that the latter record is always greater than former one. If that's the case, maybe we should reword the step to reflect that.

Reply:
+1

Reply:
I will update the step wording to reflect that we only care about the increase amount from the first to second sample

Reply:
See latest commit

Reply:
Rename the step to one of these as discussed offline: 

`Then the CPU usage <metric1> is greater than <metric2> by atmost <threshold> <units>`
`Then the CPU usage <metric1> is no greater than <metric2> by <threshold> <units>`
`Then the increase in the CPU usage from <metric1> to <metric2> is less than <threshold> <units>`

Reply:
Updated in latest commit, also added changes to logging.